### PR TITLE
BUGFIX: Migration to correct order of removed node events

### LINKED ISSUE #5352: BUG: Structure adjustments "repaired" orphaned Nodes which is not replayable
A replay in beta 15 might fail with this error

> Cannot update node with copy on write since no anchor point could be resolved for node d6cefdef-5aa4-4f27-9722-3f477c0d2c71 in content stream 092cd743-55dc-4c1e-b62c-9e4543704740

TLTR: we didn't correctly delete recursive child nodes in the past, which allowed events to exist that shoujldn't and cannot be replayed.

------------

Denny and christian did some debugging for this, with the following noticeable results:
- the event in question appears to have just appeared in the live content stream, it's not related to any publish event at all, has no metadata, no causation or correlation and there is a whole block of them appearing together.
- it's a `NodePropertyWasSet` on the node in question, in which a property is unset! that might be a clue at how these events happened. our thought was maybe a node migration but that should have happened in a workspace with a publish, but we don't see such events...
- the node in question  did exist correctly in live, BUT was indirectly removed by a grandparent removal, therefore no NodeWasRemoved appears anywhere yet the node is simply not there, we found the respective event removing the node but it is far before the offending event.
- how did the NodePropertyWasSet event ever get into the DB given that the node it applies to cannot have existed at the time the even was created and thus a constraint check should have prevented it.

------------

So far we can explain the problem due to a bugfix in `NodeAggregateWasRemoved` and the use of structure adjustments before:

There is one known case which emits events without metadata: structure adjustments 
Two of these use `\Neos\ContentRepository\StructureAdjustment\Adjustment\PropertyAdjustment::removeProperty` which would cause to immediately PUBLISH a`NodePropertyWasSet` without metadata and no constraint checks.

The query used for structure adjustments is `findNodeAggregatesByType`, but it behaves well if the projection is okay which does not seem to be:

We tested that a change in beta15 https://github.com/neos/neos-development-collection/pull/5320 which fixes `NodeAggregateWasRemoved` to correctly delete previously orphaned nodes (commit https://github.com/neos/neos-development-collection/pull/5320/commits/9df594891eb0d04a4df18f8cc5db356be379d64c) causes this problem to be noticed.

> This also fixes a bug in NodeRemoval where only one child node was removed per removed parent

------------

As a migration to fix this, we can introduce possibly a NoopEvent that is will be inserted at the places of the malformed events. 
To automate this (super expensive but i see no other way) we need to replay the content graph one event at a time, or as much until we get the errror? And then get the event, see if it has no metadata, and then replace it with a NoopEvent and continue the replay.

Alternatively we would need to have the graph projections logic in memory just to keep track which node was removed and which not.

_The NoopEvent could actually reduce complexity as WorkspaceWasRenamed and other legacy events that are just noops could be transformed into these as well and thus removing legacy code._

------------

That we dont set metadata or correlation infos always MUST be addressed before release:
- see https://github.com/neos/neos-development-collection/issues/3887
- wip pr: https://github.com/neos/neos-development-collection/pull/4969

-----------

See also our slack discussion: https://neos-project.slack.com/archives/C04PYL8H3/p1730799833465089

### PR DESCRIPTION
**Pull request by mhsdesign**
Resolves partially: #5364 a replay from previous betas works again
Resolves: #5352 by fixing the replay allowing users to upgrade to beta 15

```
migrateevents:reorderNodeAggregateWasRemoved
```

**Upgrade instructions**

**Review instructions**

The idea to solve the replay for both cases (when structured nodes were repaired, and when a node was moved after deletion) we could just reorder all the `NodeAggregateWasRemoved` events to the end.

I tested that replaying works again for both bugs.

The only downside is that the removed nodes would temporarily reappear in every workspace as they are now outdated and have to be rebased. But that can be automated as well :)

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by mhsdesign** [comment]
As discussed with Denny, Christian, Sebastian, Bastian and others this is _wrong_ in the sense of event sourcing as the history is changed but the best solution we have. Lets merge it and have it for beta 16!

**Comment by nezaniel** [comment]
Aye, reordering is bad, then again the events never should have happened in the previous order in the first place. So this is actually _correcting_ history.

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php [hunk 1]
```diff
@@ -736,6 +743,68 @@ public function migrateWorkspaceMetadataToWorkspaceService(\Closure $outputFn):
...
+        // get all NodeAggregateWasRemoved from the live content stream
+        $eventsToReorder = iterator_to_array($this->eventStore->load($liveContentStreamName, EventStreamFilter::create(EventTypes::create(EventType::fromString('NodeAggregateWasRemoved')))), false);
+
+        // remove all the NodeAggregateWasRemoved events at their sequenceNumbers
+        $eventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId);
+        $this->connection->beginTransaction();
+        $this->connection->executeStatement(
+            'DELETE FROM ' . $eventTableName . ' WHERE sequencenumber IN (:sequenceNumbers)',
+            [
+                'sequenceNumbers' => array_map(fn (EventEnvelope $eventEnvelope) => $eventEnvelope->sequenceNumber->value, $eventsToReorder)
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:777]: can these delete from and below the `commit` statements get tooo large? I tested it with 800 events to delete and publish at once and that worked.

**Comment by nezaniel** [comment]
Worked for my project, so +1

**Review by nezaniel — APPROVED** [review]
works for me

**Review by kitsunet — APPROVED** [review]

**Comment by mhsdesign** [comment]
We found out that there is a catch after all. Its is not correct for every case to move the removal to the end.
In the case

- create node `abc`
- remove node `abc`
- create node `abc`

the remove node MUST be in the middle position. Otherwise there will be a duplicate node entry in the content graph because we dont have constraints, and the document uri path constraint `variant` is violated.

This is not a problem with user generated content but for imports with deterministic node aggregate ids (which have been deleted and recreated) this will fail.
For now those projects must not apply this migration then. In case there is a project which still needs this migration but cannot use it we need to offer another way - which will be hard.

**Pull request merged**